### ISSUE TITLE
Reconsider handler initialization

### ISSUE THREAD

**Issue by CxRes** [solid/solid-rest#59]
Currently, handler initialization is based on checking the window object. I presume this has to do with detecting if the environment is a browser. However, this check is a problem in nwjs/electron environments where both `window` and `node` are defined. In such cases we still want to create file handlers and can even use native localstorage of the browser. At least checks should be made for both `window` and `process`.

I am not sure of the best strategy to manage this tough and would like to hear a bit more about the design intent in automatically initializing handlers. I see that one can pass options through solid-node-client (incl. `options.handlers`) but these are not documented. That is why I have refrained from putting up a PR!

Somewhat unrelated but relevant to the issue, I think in the long term solid-rest should be decoupled from solid-node-client (and other providers  should be decoupled as well, as this list will become unwieldingly large if this project takes off). Each type of source then should be a plugin that is invoked based on user-installation (perhaps done so automatically on first invocation of the scheme) and url scheme. In fact, this is the approach I am taking in including solid in my application. Please consider this in your future plans!

**Comment by jeff-zucker** [comment]
I think you have misread the code for the handlers.  (It's a bit messed up because I had to retrofit some things to maintain backwards compatibility and also handle a passed in rdflib for patch.) The way it works is that **if** the user does not pass in one or more handlers, it checks the operating environment and pulls in the default handlers for that environment.  However, if the user does pass in some handlers, it uses those and does not check the environment at all.  So in Data-Kitcehn (an electron app), I pass in both the nodejs and the browser-based handlers in the call to new SolidRest() because it is capable of operating in both environments.  

I am behind on documenting Solid-Rest.  I need to do a lot of work there and mentioning options.handlers is a big one.

I am unclear what you mean about decoupling from solid-node-client.  There is nothing in solid-rest that requires solid-node-client and solid-rest is quite capable of operating without it.  And in solid-node-client (although undocumented as yet, I just put this feature in) you can pass in any fetch handler with the call to new SolidNodeClient.  Additionally, rdflib currently looks for global.solidFetcher allowing users to define the fetcher at the top of their app rather than passing it in when creating a fetcher.  Solid-client-authn-nodejs (if it ever comes out) also plans to look for a global fetch to allow users to override the built-in.  So it seems to me that an app can easily mix and match  auth/login libraries and fetch libraries.  Am I missing something?

If I've misunderstood your suggestions or my response doesn't answer them, please ask again.

**Comment by CxRes** [comment]
>  The way it works is that if the user does not pass in one or more handlers, it checks the operating environment and pulls in the default handlers for that environment. 

I think this could do with some improvement (since it fails in nwjs), by checking for `node` and `process` objects for instance. Just because window is undefined is no reason node.fs is defined or vice-versa. Check should be something on the lines of: Is node defined? then FS; Is localStorage defined? then localStorage else Memory. 

In my case of nwjs, it goes `false` on `typeof 'window' === undefined` , then fails the `try` because `browserFS` is not a constructor (for reasons I do not understand as of now) and ultimately crashes at `handler.forEach` because `handlers` is initialized as `null`! I am guessing, the bug is that `handlers` should at least be an empty array. But we came to that only because we could not define any handler at all!

Meanwhile could you please point me to the handler invocation in Data-Kitchen. At least, I can side-step the issue that way for now.

> There is nothing in solid-rest that requires solid-node-client.
Its the other way around that I have an issue with, ie solid-node-client requires solid-rest. So just to use solid-node-client (essentially the auth and fetch in it with say, solid-file-client) I have to pull in all of solid-rest. Let me put this in another way:
In my ideal scenario (this is my opinion, though held strongly), solid-rest would be a thin wrapper that passes controls to different protocols and one would install solid-node-client or solid-localstorage etc. as plugins. Currently, fs: or app:/ls/ etc. come baked in with solid-rest (even if I do not want to support one of these protocols in my end product or any for that matter) which in turn come baked in when I install solid-node-client. It is a classic banana-gorilla problem! 

If there is another mechanism you have in mind or currently exists that I am missing, I am all ears!!!!!

> Additionally, rdflib currently...
I'll come to that in a separate issue. rdflib uses outdated deps and I have had Rollup screaming madly at me last night when I activate it. I hope we can find some other way to enable PATCH.

**Comment by jeff-zucker** [comment]
Solid-Node-Client does nothing but add solid-rest's fetch to whatever  the underlying auth is.  Currently that is solid-auth-fetcher but if and when there is a solid-client-authn-node, it will switch to that.  If one doesn't need solid-rest, one can use either of those auth libraries directly without solid-node-client.

The failure in nwjs (which I had never heard of but looks interesting) is concerning.  My feeling is that I should add a note in the documentation that if you are operating in a mixed environment such as nwjs or electron, you MUST supply your own handler(s) in the call to new SolidRest(), see below for an example.

I agree that solid-rest should be as light as possible, but for backward compatibility and general principles, I would like it to function out of the box for file:// requests in node.  Maybe I'm wrong about how conditional requires work, but my impression is that file.js and the other plugins are only imported automatically if no other handlers are supplied.

Rdflib is in rapid development and so problems with including it may be temporary.  I encourage you to raise an issue or go to the rdflib chatroom if you are having diffuculty with it.  

For cases in which the user does not want the built-in plugins, passing in a handler will override them and they will not be loaded.  For example, this instantiation would use localStorage.js but not file.js.
```javascript
const SolidRest = require('solid-rest')                                         
const LS = require('solid-rest/src/localStorage')                               
const rest = new SolidRest({                                                    
  handlers : [new LS()]                                                         
}); 
```

There is a working example of doing this in a browser at solid-rest/tests/browser-test.html.  It instantiates the object like so:
```html
<script src="../bundles/browserfs.min.js"></script>
<script src="../src/browserFS.js"></script>  
<script src="../dist/browser/solid-rest.js"></script>
<script>
const rest = new SolidRest({
  handlers : [new SolidBrowserFS()]
})
...
```

**Comment by jeff-zucker** [comment]
I do not understand "install solid-node-client ... as plugins".  Solid-node-client is not a plugin to solid-rest, it's the other way around.  Here is the chain as I see it:
```
define a non-http-fetcher (solid-rest+plugins)                                  
define an http-fetcher (solid-auth-fetcher)                                     
create a hybrid fetcher that uses both (solid-node-client)                      
specify another library's fetcher as that hybrid (rdlib,solid-file-client,etc.) 
```

**Comment by CxRes** [comment]
Sorry, this post goes off topic into the plugin issue. I'll return to it in the next comment...

> Maybe I'm wrong about how conditional requires work, but my impression is that file.js and the other plugins are only imported automatically if no other handlers are supplied.

That is only true if you are invoking the code from node (and then too the code exists in node_modules). Bundlers like Webpack and Rollup will pull all the requires unless you specifically ask it not to (which requires you to dig in and manually identify these require calls). I have all this code in my bundle even when I do not want to use most of these fetchers. I would like to avoid that!

> I do not understand "install solid-node-client ... as plugins".

That clarifies my mental model a bit (perhaps influenced by solid-auth-cli, or my impression of it). Hopefully, I can now explain my idea better.

In my model, I would ideally like to see a fetcher manager, lets call it solid-fetch-manager wherein http-fetcher, file-fetcher, localstorage-fetcher etc. each will be a plugin. Now, I imagine 2 modes:
1) Manual mode: wherein developer will choose what fetcher they want. (It is not inconceivable if solid catches on, a developer may create an app that only stores data locally for a session using say, a local-storage fetcher (and not even use a http-fetcher). Or another protocol instead of http catches on? Or I want a desktop only app! Ultimately, it is solid that matters as the standard for data communication not the transport/storage, that it is built first on http is incidental. *And solid-rest is proof of that imho!*)
2) Automatic mode: where based on the protocol on the address, the fetcher-manager installs or downloads the correct fetcher from a CDN. (would work extremely well with something like deno)

Ultimately, I am trying avoid any bloat being added due presumptions made, even if it is a completely reasonable presumption like supporting http: and file: out of the box. To support these assumptions, we can simply install http: and file: fetchers anytime one installs the solid-fetch-manager.

Re: Example
Just one pedantic point which drags on the plugin issue... it is generally a bad idea in my experience to call files inside a library ie `solid-rest/src/localStorage` (since they might get remapped/renamed at some point, I have filed such bugs before). Partially, we can fix this with the exports property in package.json, but really, I think it is a sign that they need to be in separate repos!

**Comment by CxRes** [comment]
NWjs (https://nwjs.io) was the original electron. One of their alumni was responsible for much of the work on electron (and it got pretty acrimonious). It has both `window` and node environments together. Thats why I am proposing a better check model for handler association. I can make a PR for you to review, since I think it would be a non-breaking change otherwise!!!

**Comment by jeff-zucker** [comment]
I am thinking of this as the interface for SolidNodeClient. In this example, the user does their own importing and merely passes things down the line.         
```javascript                                                                   
let client  = new SolidNodeClient({                                             
  fetchers : [                                                                  
    { http : new SolidAuthFetcher() },                                          
    { file : new SolidRestFile() },                                             
    { bfs  : new SolidRestBrowserFS() },                                        
  ],                                                                            
  parser : require('rdflib'),                                                   
})                                                                              
```                                                                             
The fetchers are optional - user can import only ones for protocols they expect and the parser is optional, only required if user wants PATCH.  SolidRest would use only the pre-imported handlers and parsers passed to it, and would not directly import any of the plugins or rdflib so would not have any of them in the dependency chain.                                                               
                                                                                
If new SolidNodeClient() is called with no fetchers, it would default as now.  This would mean that it would need to import SolidRestFile. But SolidRest would not have to import it.  So if the user wants SolidRestFile out of the dependency chain, they can just use Solid-Rest directly.                                
                                                                                
How does that sound?

**Comment by CxRes** [comment]
> How does that sound?

This sounds much closer to what I had in mind!!!!!

> If new SolidNodeClient() is called with no fetchers, it would default as now. This would mean that it would need to import SolidRestFile. But SolidRest would not have to import it. So if the user wants SolidRestFile out of the dependency chain, they can just use Solid-Rest directly.

This confuses me (perhaps I am missing something about the internal design). Why would one need solid-rest anymore (except for legacy reasons) when doing this? Couldn't one design it in such a way as to...
```js
import { SolidNodeClient } from 'solid-node-client';
import SolidAuthFetcher from 'solid-auth-fetcher';
import SolidRestFile from 'solid-rest-file';
import SolidRestBrowserFs from 'solid-rest-browser-fs';

let client  = new SolidNodeClient({                                             
  fetchers : [                                                                  
    { http : new SolidAuthFetcher() },                                          
    { file : new SolidRestFile() },                                             
    { bfs  : new SolidRestBrowserFS() },                                        
  ],                                                                            
  parser : require('rdflib'),                                                   
}) 
```

Am I right in presuming currently you plan to:
```js
...
import SolidAuthFetcher from 'solid-auth-fetcher';
import SolidRestFile from 'solid-rest/src/file.js';
import SolidRestBrowserFs from 'solid-rest/src/browser-fs.js';
...
```

AFAICT, there is nothing except organization of repos that privlidges http fetcher over non-http fetchers? Or again am I missing something? 

There is one other thing that is bothering me but I am afraid I do not have clarity of mind about the problem itself. My concern is that on local networks, such as the one I have at home, one has a login per computer (which effectively is a domain, though not officially). So, I have different logins for `file://computer1/D:/path/to/file` and `file://computer2/path/to/another/file` (`file:` with two slashes and not regular 3, like `http://`) which is different for my local computer `file:///yet/another/path`. My concern is that if I have to mix and match data from these fetch calls we would need to have separate clients, even though each login pertains to each individual computer. Similar issues would arise in `http:` when user has multiple identities but wants fetch data from different domains (though here the common login might work across domains) towards a common task. I do not know what is the best way to address this and would like more time to think this through. But I just want to put this concern out there...

**Comment by jeff-zucker** [comment]
> Why would one need solid-rest anymore?  

Because it is the glue that makes the plugins work.  SolidRestFile would be the same as the current solid-rest/src/file and, like it, would need solid-rest to call it.

The local network is an interesting issue,  I haven't thought about it.  But there would be no login - solid-rest and solid-node-client do not have any login for local files, only for pods.  I don't see why different clients would be needed.  Can you do a test and see if you can do a fetch against each of the URLs you mentioned?

**Comment by CxRes** [comment]
> Because it is the glue that makes the plugins work

I thought solid-node-client is the glue.  It seems strange that you need separate glue first for non-http and then tie in http with the non-http. Should any and every protocol not be at par?

> Can you do a test

I can tell you as it is it would not work without OS specific authentication (at least for windows). And I am not worried if it does not work just today, but the provision for long term!

**Comment by jeff-zucker** [comment]
Solid-Node-Client creates a client that calls an auth package to perform login and authenticated fetches for http requests and calls solid-rest to handle non-http requests.  It only works in nodejs, not in a browser.  It does not do any heavy lifting, it just creates a hybrid fetcher that performs both kinds of fetching.  It does not understand or interpret either requests or responses, merely passes them on.

Solid-Rest works either in nodejs or in a browser, depending on which of its plugins is used.  It gets requests for non-http resources from a client  (SolidNodeClient, or directly in an app,etc.) and, depending on the protocol, hands the actual interaction to a plugin such as solid-rest/src/file, gets back a response, packages that as a Solid HTTP response and sends it back to the client.

Anything specific to a nodejs environment belongs in Solid-Node-Client or in a Solid-Rest plugin but not in Solid-Rest itself.

I do not at all understand by what you mean by "at par".

I also do not understand what part of your networking example would need to be implemented in solid-rest or how solid-rest would need to behave differently to accommodate a plugin oriented to  a networked environment.  It seems to me that you could send an ssh login then proceed to use solid-rest over the ssh connection.

**Comment by CxRes** [comment]
Loooooooong post warning...

> At Par

Indulge me here... Solid is meant to be a universal linked file system, that is, it does not matter where the data is and how we access it. We give Solid an address, it gives us our data and if that data contains links, it goes back and fetches that data as well. It does not matter if that address begins with http or file etc because http, file etc are merely physical storage/transport. This we already agree upon. 

Now the current architecture looks like this:
```
client (solid-node-client)
|- http fetcher (solid-auth-fetcher)
`- non-http fetcher (solid-rest)
    |- file fetcher
    |- ... etc.
```

The fact that you switch in two steps mean that you consider http to be privileged. From what I have understood, this privilege is due to the fact:
1) Solid was designed for the web first and on top of http.
2) On the web authentication is a must (for editing). 

This special status to http is an implicit assumption in how things stand. And this assumption violates the generality that Solid strives for or should strive for. This in the future will become a source of complexity (in that, you will forever maintain atleast two switching layers). 

A different way to design would be:
```
client 
|- http fetcher
|- file fetcher
|- ... etc.
```

Apart from maintaining a single layer, it also means that other protocols will be first class citizens. Think about this: if a computer in the future boots into Solid os, the first thing you would need to do is authenticate user on the file: protocol. You will far more often access file: over other protocols. Now, for example, is it not inefficient to first decide http or non-http and then go to file.  

I see that there is a browser/node dichotomy, which creates the need for separate fetchers on the same protocol. On one hand browsers prevent file system access on the other hand node does not have native fetch. But both these things will go away. FS is coming to the browser and Deno has native fetch. My concern is that practicalities of today will lock in the design for many years with all that extra complexity. 

Eventually we should strive to hide data storage itself from the user and for that Solid will have to be universal. And let me stress again that solid-rest is the best proof of this (and perhaps why I am eating your ear off!). Stated differently, solid-auth-client, solid-auth-cli or solid-auth-fetcher all should also just be clients of solid-rest since these are only just handlers. 

Please just give it a little time, I think you will also come around to see beauty in the symmetry, I promise!!!!!!! 

> Networking example
For many applications ssh is an over-kill. Eg fetching some data from a NAS. Having said that I am going to reiterate that I lack clarity myself, so it is difficult to say much more for now!!!

**Comment by jeff-zucker** [comment]
I think this may solve your issue with being "at par" : 
```javascript
const client = new SolidNodeClient({                                            
   authHandler :    "",       // solid-auth-fetcher/solid-client-authn-node/...    
   protocolHandlers : {                                                         
     http : "",               // node-fetch/cross-fetch/...                        
     file :   "".             // solid-rest-file/...                               
     ...                                                                        
   }                                                                            
});                                                                             
```
Plain http is at par with the other protocols but the authHandler is privileged, not because it uses http, but because it provides login and authenticated services.

**Comment by jeff-zucker** [comment]
Or, to consider that there may be authHandlers for other protocols these would be the defaults :
```javascript
new SolidNodeClient ({                                                          
  authHandlers : [                                                              
    { http : 'solid-auth-fetcher' },                                            
  ] ,                                                                            
  noAuthHandlers : [                                                            
    { http : 'node-fetch' },                                                    
    { file : 'solid-rest' },                                                    
  ], 
  parser : ""                                                                            
}) 
```

**Comment by CxRes** [comment]
As I mentioned above, a file system access requires authentication. As solid-rest matures, it is inconceivable that file system authentication will not be implemented despite the complex engineering needed to deal with the way file systems are designed at present. So you will end up with a situation that in your example will look like:
```js
new SolidNodeClient ({                                                          
  authHandlers : [                                                              
    { http : 'solid-auth-fetcher' },
    { file: 'solid-rest-file-with-auth' }                                            
  ] ,                                                                            
  noAuthHandlers : [                                                            
    { http : 'node-fetch' },                                                    
    { file : 'solid-rest-file' }, // I think you meant solid-rest-file and not solid-rest here                                                    
  ], 
  parser : ""                                                                            
}) 
```

I would argue that you might as well get rid of this authenticated and non-authenticated handler distinction altogether. All authenticated handlers can access data from sources that do not need authentication. So we just end up with:
```js
new SolidNodeClient ({                                                          
  handlers : [                                                              
    { http : 'solid-auth-fetcher' },
    { file: 'solid-rest-file-with-auth' }                                            
  ] ,                                                                            
  parser : ""                                                                            
}) 
```

I can see that there shall be handlers where authentication is meaningless, such as, localstorage. But such a handler is a special case of a handler with authentication, where the authentication is always `true`. 

Close your eyes and imagine a computer that starts only with a solid browser only. What do you think is the best way to manage user identification might be across all resources, local or remote? How do we minimize user and programmer burden in using solid? You may have different answers, most probably better answers, but we need to have this conversation before too much gets baked in!

**Comment by jeff-zucker** [comment]
Well that's all true but leaves out the case where we do unauthenticated http fetches - which should happen automatically if the user is not logged in but which are replaced with authenticated fetches when the user is logged in.  So we need to be able to specify both.  How does this look?
![solid-node-client](https://user-images.githubusercontent.com/44732708/101291612-ed54be80-37be-11eb-8b43-85da0bc2bb0b.png)

**Comment by jeff-zucker** [comment]
There could be authenticated fetches for file:// or other protocols, but presumably those would also do unathenticated fetches so there is no reason to include two paths.

**Comment by CxRes** [comment]
**I really like the chart, especially the bit where users can actually switch out handlers for the same protocol, which could be used to support different environments or performance characteristics.** Just some confusion regarding http (that needs some clarification from you, see below).

> Well that's all true but leaves out the case where we do unauthenticated http fetches - which should happen automatically if the user is not logged in but which are replaced with authenticated fetches when the user is logged in. 

I did not follow. Could you please elabourate! Also I am confused by the second comment...

I would presume that you can always "fetch", whether you are logged in or not. The difference would be that you either send credentials in your request headers or don't. If you did not log in and placed a fetch, a response will still come, it just may not contain the data you are after. If you logged in later then you will just have to place a new fetch. 

Once solid-auth-client has obtained its cookie (or another auth strategy), it also uses a fetch library underneath. Now that I think of it, it may be wise to just separate all auth logic from the fetch logic (though auth will itself use fetch to get identity from a server and feed that to subsequent fetches). Similar thing in terms of separation, though not the same, might be valid for the file-system as well (login the user using one code ie auth and then read data using another ie fetch). So this different model would look like:
```js
new SolidNodeClient ({                                                          
  authHandlers : [                                                              
    { http : 'solid-auth-getter' },
    { file: 'solid-rest-system-auth' }                                            
  ] ,                                                                            
  fetchHandlers : [                                                            
    { http : 'solid-auth-fetcher' },                                                    
    { file : 'solid-rest-file' },                             
  ], 
  parser : ""                                                                            
}) 
```

But, I need to know first whats on you mind with your last two comments....

**Comment by jeff-zucker** [comment]
Yes, I think I am over-thinking it.  I will remove the authenticated/unauthenticated distinction.   Each handler for a protocol could do both auth/noAuth or only noAuth at its discretion.

**Comment by jeff-zucker** [comment]
![solid-node-client2](https://user-images.githubusercontent.com/44732708/101293941-e92d9e80-37c8-11eb-9870-ac925dba2f20.png)

**Comment by CxRes** [comment]
Looks good! 

I'll try to submit a PR sometime this week wrt automatic initialization (though not necessary, i think should help the unsuspecting user). 

Perhaps, you may consider splitting solid-rest to separate plugin repos at some stage!

**Comment by jeff-zucker** [comment]
Thanks for helping me think this through, I like the end result a lot.

**Comment by jeff-zucker** [comment]
Okay, so I've taken a stab at a plugin system, seems to work but maybe not the right way.  Please take a look.  See the README on plugins.  https://github.com/solid/solid-node-client/tree/solid-client-authn

**Comment by CxRes** [comment]
You would need to explain some things:
* what is ESS and CSS?
* why is there a separate cookie fetching code. Could that not be integrated within `solid-client-authn-node`?

Also, I think fallback should not be "node-fetch" but a piece of code that simply throws "no handlers associated with protocol foo:", like with the empty string case. Perhaps `node-fetch` might do that anyway as well as cover `http:` but, its confusing in that case and should be explained in documentation.

**Comment by jeff-zucker** [comment]
ESS is Inrupt's Enterprise Server.  CSS is Community Solid Server - Ruben Verborgh's project that is not as feature rich but more spec compliant than NSS and is planned to replace NSS.

There is separate cookie fetching code because neither ESS nor CSS have or plan to have a command-line based method to obtain the token/secret.  Getting the token/secret via a browser is the only option.  I have spoken to Inrupt on this and they will not budge.  So all I am doing is pointing people to Inrupt's bootstrap script (I also submitted a PR to  Inrupt to improve that script).

I hear you on the fallback, I still think it is an appropriate name because someone may want to provide their own fallback library.  There's no reason for us to check if a protocol exists, node-fetch will error on its own.  For sure the docs could use some work.  I've revised the README this morning plus mentioned you in the acknowledgements.  Thanks again!

**Comment by jeff-zucker** [comment]
It is a fallback for both http: and unauthenticated https:.  Among other reasons, it lets us not load the authenticated handler unless we need it but still work for https: requests.

**Comment by CxRes** [comment]
> Getting the token/secret via a browser is the only option.

Isn't it still ultimately a https: request? At worst we could `import` the Inrupt cookie fetcher in https: handler internally and hide that from the user, maybe even persist it on disk.

> Getting the token/secret via a browser is the only option. I have spoken to Inrupt on this and they will not budge.

Do you know whats going on here? Is it that they have a different idea of Solid than we do? Or is it some security thing? How does it matter what environment fetches Solid data?

The thing I hate about solid-auth-client is that actually does a redirect when accessing private parts of the pod, and opens the login window with a webpage. Ux should always be independent of the back-end negotiation and at discretion of the user/developer. 

> I still think it is an appropriate name because someone may want to provide their own fallback library.

> It is a fallback for both http: and unauthenticated https:. 

* My concern is that fallback is doing too much (And the person providing an alternate fallback library will have to cover these special cases). 
* Unauthenticated https: fetches should be handled by same code as authenticated https: code (internally it is calling node-fetch or cross-fetch anyway)! You could still do `client.fetch(/*https: address*/)` without `client.login(/*https: address*/)` and it still uses https: handler!
* http + other protocols with node-fetch as fallback is OK though for now! (We could still have public pod accessible through http, if the user wants to use another handler for that than node-fetch, that handler itself must also have throw logic for undefined and non-http protocols. I favour separation because it seems cleaner to explain to users.  Even something like this is more intuitive:
```js
    const client = new SolidNodeCient({ handlers : {
      http : 'alt-node-fetch', 
      fallback : 'node-fetch',          // all other fetches
    }})
```
But it does not bother except as a detail me if we have good docs.)

**Comment by jeff-zucker** [comment]
The token/secret is a security issue.  It works this way in almost anyplace that requires you to get an store a secret or api-key.

They just released it, I am certainly not going to start building wrappers around it before they get further along.

Do not use solid-auth-client it is deprecated.  Use Inrupt's solid-client-authn-browser instead.  It does not support any kind of redirects but again, it is a security issue, and wiser people than me have tried to talk them out of it.

The current setup includes the possibility of defining an http handler.  If someone wants that , they can just create it.  That's a separate question of having a fallback.  It's simply that, using the current defaults nothing handles http: so it goes to fallback.

**Comment by jeff-zucker** [comment]
If I have conditional requires, as I do in src/getAuthSession.ts, and I include those packages as devDependencies, not Dependencies, then downstream apps could import solid-node-client without those if it wanted (e.g. couldn't handle 'fs')  but someone who npm installs the package gets all of the plugins installed and can use them out of the box.  Maybe?

**Comment by jeff-zucker** [comment]
Also mark them as PeerDependencies.

**Comment by CxRes** [comment]
I have concerns with how you are coding:
* Either use esm or cjs but don't mix both. `import()` can be used in place of `require()`.
* solid-node-client will be a peer dependency of solid-rest-file, if solid-rest-file were its own npm package.
* solid-rest-file will be an optional dependency of solid-node-client, in that case. If a downstream app includes through npm solid-node-client, it will pull in solid-rest-file (npm I think always tries to install optional sub-dependencies, except that it ignores failures for optional dependencies). This is inevitable unless you get rid of default setup or put the default setup in a separate npm package (which would not be a bad idea). Only, we can build solid-node-client in such a way, that if we import it in from a CDN, it'll will only conditionally import solid-rest-file from the CDN.    
* Do not use devDependencies for this. They are not for this use case!

**Comment by jeff-zucker** [comment]
> solid-node-client will be a peer dependency of solid-rest-file,

I do not understand that at  all.  Solid-Node-Client instantiates a series of fetchers.  Solid-Rest-File is one of those fetchers.  SNC should import Solid-Rest-File, not the other way around.  How would that work?

> import() can be used in place of require()

AFAIK, not if the require should be conditional.  Is conditional import possible?  If not, how do we write Solid-Node-Client?

**Comment by CxRes** [comment]
Peer deps essentially say that if you want to use me, you need to fetch my host package. Read here: https://nodejs.org/en/blog/npm/peer-dependencies/

So if someone installs SRF, they need to get SNC (or a warning to get SNC). SNC on the other hand defines SRF as optional dep (because one can use SNC without SRF, if they know what they are doing). 

Conditional Imports is very much possible and one of the use cases for `Import()` See: https://v8.dev/features/dynamic-import or other resources on the subject.

**Comment by jeff-zucker** [comment]
> So if someone installs SRF, they need to get SNC (or a warning to get SNC). SNC on the other hand defines SRF as optional dep (because one can use SNC without SRF, if they know what they are doing).

Solid-Rest-File is a plugin for Solid-Rest.  It is not meant to do anything at all on its own. There is no circumstance in which someone would want to install it on its own.  It has no public user interface, only an interface that is called by Solid-Rest. 

I'll read the docs on import, thanks.

**Comment by CxRes** [comment]
Can you clarify what is the final relationship between solid-rest and solid-node-client. I thought solid-node-client will call solid-rest with node-specific plugins and solid-rest will be requiring fully manual initialization?

**Comment by jeff-zucker** [comment]
Solid-Rest, by default, will load no plugins except those passed by the user or app that calls new SolidRest( [plugins] ).  So if someone does not want fs and the other dependencies of solid-rest-file, they can load SolidRest without solid-rest-file.

Solid-Node-Client will, by default, instantiate a new Solid-Rest object with the Solid-Rest-File plugin and use that to handle file: requests. 

Does that answer your question?  Do you have a different way it should work?

**Comment by CxRes** [comment]
Ok! So you can then merely substitute solid-node-client with solid-rest wrt what I said about peer deps! Solid-rest-file etc. will be proper deps of solid-node-client. Does that sound right now?

**Comment by jeff-zucker** [comment]
Solid-Rest and Solid-Rest-File would both be dependencies of Solid-Node-Client.
Solid-Rest-File would be an optional plugin for Solid-Rest, and not a dependency of it.

Does that sound right?

**Comment by CxRes** [comment]
Yes. And solid-rest is a peer dep of solid-rest-file!

**Comment by CxRes** [comment]
Also, I am closing this with #61.

**Comment by jeff-zucker** [comment]
Well, #61 is not going to be implemented. I will first re-write solid-node-client using dynamic imports.  Then I will re-write solid-rest and make the plugins independent packages and remove the loading of defatult plugins and the environment checks.

**Comment by jeff-zucker** [comment]
Here is the problem with dynamic import : it requires async, so can't be called from constructor so new SolidNodeClient( some_plugin ) will not be able to use it.  I guess I could make a installPlugin() which users would have to call after the call to new.

**Comment by jeff-zucker** [comment]
Or maybe this would work : only load the auth package on login - the package isn't needed before login, so there is no reason to load it before.

**Comment by CxRes** [comment]
Its ok, just that fix is available, even if this code goes into a maintenance branch subsequent a rewrite. You can accept it there or not, for history sake its there!

**Comment by CxRes** [comment]
Packages may be needed before login, but handler initialization/client creation can always be async!

**Comment by jeff-zucker** [comment]
How can it be async in a constructor?

**Comment by jeff-zucker** [comment]
Yes, right, node-fetch and solid-rest would be loaded automatically, only the auth packages would be delayed until login.

**Comment by CxRes** [comment]
You can do (internally or externally):
```js
async function addHandler({protocol:handler})
```

user can or you can internally
```js
client.addHandler(...)
```

**Comment by jeff-zucker** [comment]
Those can be done after the constructor but would not be backward compatible because users expect to be able to do login without calling addHandler() or whatever.

**Comment by CxRes** [comment]
I would import all the code at handler initialization and not wait till the user calls login.

**Comment by CxRes** [comment]
You can choose not to expose addHandler, if you like...

**Comment by jeff-zucker** [comment]
So import both solid-auth-fetcher and solid-client-authn-node on call to new?

**Comment by jeff-zucker** [comment]
Neither of those packages can be called before login.

**Comment by CxRes** [comment]
> So import both solid-auth-fetcher and solid-client-authn-node on call to new?

Otherwise login will have to wait on the cdn which is waste of time! Fetch everything in parallel at the first chance.

**Comment by jeff-zucker** [comment]
So, in other words, no conditional import.

**Comment by jeff-zucker** [comment]
Because async addHandler() can't be called in the constructor.

**Comment by CxRes** [comment]
Its still conditional import! You only import if the handler is invoked. Just the conditional import is at client construction time.

Though, on second though, it may actually be more performant not to conditionally import default handlers.

**Comment by jeff-zucker** [comment]
> conditional import is at client construction time.

How is that possible if conditional import requires async and constructor forbids async?

**Comment by CxRes** [comment]
You can always do:
```
const client = new SNC();
async client.init();
```

though see above.

**Comment by jeff-zucker** [comment]
Yes, I KNOW I can add a new call, but that breaks backwards compatibility,

**Comment by jeff-zucker** [comment]
I guess I will auto-load solid-rest, node-fetch, and solid-auth-fetcher which will give the existing behavior.  If users wants to alter the handlers they can call await addHandlers() which would dynamically load other  packages.

**Comment by jeff-zucker** [comment]
But that is ugly because they will need to specify protocol handlers in both the new() and addHandler().

**Comment by CxRes** [comment]
Yes, I see. Thats why I don't like classes and prototype inheritance in JS! 

Your solution looks OK for now.

Let me see if I can come up with something more satisfactory. But, please don't revert to mixing cjs and esm. Its just terrible for downstream consumers!

**Comment by jeff-zucker** [comment]
Yeah, I hear that, eliminating the cjs now.

**Comment by jeff-zucker** [comment]
Ugh, I just remembered that solid-auth-fetcher and solid-client-authn-node (which are two different branches of the same code) clobber each other if loaded together in the same process.  So I MUST ensure that only one of them is loaded.

**Comment by jeff-zucker** [comment]
I think delaying loading auth until login and then doing it conditionally is the way to go.  It would only happen once and would not take any more time than await addHandler() would.

**Comment by CxRes** [comment]
Wait a minute. This is not really a problem for solid-node-client. Wouldn't you expect a user to import a non-default handler first? How they do it is their business!

**Comment by CxRes** [comment]
It is only a problem for backwards compatibility, in that you cannot just feed a string as a handler, in non-default cases. I think that should be acceptable though.

**Comment by jeff-zucker** [comment]
Even if we eliminate a string as a handler, we have the clobber problem.  We load solid-auth-fetcher by default, then use imports and instantiates with solid-cient-authn, they clobber.

**Comment by jeff-zucker** [comment]
Am I correct that doing a conditional import in login() as oppossed to addHandler() takes the same amount of time?  What is wrong with doing it in the login?

**Comment by CxRes** [comment]
That is from the new code base, so I don't know much abt it!

**Comment by jeff-zucker** [comment]
And "CDN" is not in play - this is a nodejs app.

**Comment by CxRes** [comment]
Don't do conditional imports at all. Let the user import!

**Comment by jeff-zucker** [comment]
In other words screw everyone who used the package previously?

**Comment by CxRes** [comment]
Why does it screw anyone at all?

**Comment by jeff-zucker** [comment]
because previously client = new SolidNodeClient(); client.login(); just worked!

**Comment by jeff-zucker** [comment]
And I do not want the user to have to import an auth package, solid-rest, solid-rest-file, and node-fetch before using Solid-Node-Client

**Comment by CxRes** [comment]
It will still work!

But I see what you mean. I am mixing up two things - you still want to conditionally load. But isn't it the case that in the default handlers case the require always ran.

**Comment by jeff-zucker** [comment]
> it can still work

No, if the user must supply the auth package, a call to new with no arguments followed by login will not work.

The current getAuthSession.ts does conditional requires based on what the user supplies.  If I change those to imports, I can't call it from new()

**Comment by CxRes** [comment]
Ok, the way around it would be to run an effective `client.init()` on the first of `.fetch`, `.login`, `.logout` on a per protocol basis. Something like:
```js
Class SolidNodeClient {
  ...
  async fetch(address, ...args) {
     const protocol = getProtocolFromAddress(address)
     if (!handler[protocol]) {
       handler[protocol] = await import( getHandlerLocation(protocol) )
     }
     return handler[protocol].fetch(address, ...args)
  }
```

I know its ugly, but it works within our constraints!

Though consider deprecating synchronous api in favour of an async api for a 2.0.

**Comment by jeff-zucker** [comment]
Yes. that goes in the login.  One can't do an authenticated fetch or a logout prior to being logged in, so no reason to call it there.

**Comment by CxRes** [comment]
One can do an unauthenticated fetch without a login (which will still be handled by the same handler, think fs with or without login). And if the user does a logout and no function exists, it must not throw, so its safe to do for all three. Its just one line check anyway!

**Comment by jeff-zucker** [comment]
If one is not logged in, one uses a NoAuthSession which does nothing but succeeds on logout and does a node-fetch fetch for fetches

**Comment by CxRes** [comment]
So what would the api of a custom handler be?

**Comment by jeff-zucker** [comment]
This is a NoAuthSession object.  It gets instantiated with whatever Solid-Node-Client's file and fallback handlers are, node-fetch and solid-rest-file by default : 
```javascript                                                                              
export class NoAuthSession {                                                    
  session:any                                                                   
  fileHandler: any;                                                             
  httpHandler: any;                                                             
  constructor(options:any={}){                                                  
    this.session = {                                                            
      info : { isLoggedIn:false },                                              
      fetch : this.fetch.bind(this),                                            
      logout : ()=>{},                                                          
    }                                                                           
    this.fileHandler=options.fileHandler;                                       
    this.httpHandler=options.httpHandler;                                       
  }                                                                             
  createServerlessPod( base ) {                                                 
    return this.rest.createServerlessPod(base);                                 
  }                                                                             
  fetch(url:string,options:any) {                                               
    if( url.startsWith('http') ) return this.httpHandler.fetch(url,options)     
    return this.fileHandler.fetch(url,options);                                 
  }                                                                             
}     
```
P.S. I need to learn enough about Typescript to replace "any" above.

**Comment by jeff-zucker** [comment]
The handlers could be anything with a fetch() method.

**Comment by CxRes** [comment]
Wouldn't auth be on a protocol basis as well (fs auth will be different from https auth). Not only would handlers have a `fetch()` method, they will also need their own login/logout methods. 

Only say if login/logout are absent, we presume that fetches are always unauthenticated.

**Comment by jeff-zucker** [comment]
If someone has an authenticated login for file protocol, they would set that as the file protocol handler.  Prior to login, they would fallback to the NoAuthSession and would not expect a login.  Once logged in, they would not ever get t o the NoAuthSession since it is only a fallback for when something else can't handle it, and the file: protocol will already have said it can handle it.

**Comment by CxRes** [comment]
> If someone has an authenticated login for file protocol, they would set that as the file protocol handler.

I don't follow, maybe a code example, please?

**Comment by jeff-zucker** [comment]
A code example for a theoretical package noone has written?

**Comment by jeff-zucker** [comment]
c = new SolidFileClient({
  file : new myFileThing();
})
// c.fetch('file://foo") here will use the fallback i.e. solid-rest-file for unauthenticated fetch
// c.login(); c.fetch('file://foo') will use the authenticated fetch in myFileThing()

**Comment by CxRes** [comment]
I meant some psuedo-code for how file login might work for user from solid-node-client.

**Comment by CxRes** [comment]
`c.login({file:cred})` will need some credentials and then if you have `var myFileThing = new MyFileThing()`, I thought `myfilething.login(filecred)` will log you in. Are you intending to use solid credentials for everything?

**Comment by jeff-zucker** [comment]
I think maybe such a package would have to provide its own fileLogin() method which the user would call because they would want to call login() for https servers and possible both in one script.  So the fileLogin() method would obviously take whatever credentials it wants to accept.  Already the credentials for an NSS and a CSS login are different and those are handled in the NssAuthSession and EssAuthSession packages.  File would be no different.

**Comment by jeff-zucker** [comment]
BTW, the idea of a file login makes no sense to me.  The user logs into their computer and the uses Solid-Node-Client as that logged in user.  Why would they want or need to log in again?  Currently the user's perms are recorded in the wac-allow header so if on *nix and the user has read but not write perms on the local file system, the wac-allow header will be user="Read",public="Read".  What would a login system add?

**Comment by CxRes** [comment]
But how will the user access `fileLogin()`

Either they:
```js
var fileHandler = new myFileThing();
fileHandler.fileLogin(fileCred);

c = new SolidFileClient({
   file : fileHandler
})
```

or

```js
var fileHandler = new myFileThing();

c = new SolidFileClient({
   file : fileHandler
})

c.login({
  file: fileCred
})
```

**Comment by jeff-zucker** [comment]
Why is this important now?

But I would prefer the second way except instead of c.login(file:fileCred) it would be c.fileLogin(fileCred).

**Comment by jeff-zucker** [comment]
Different clients migh have different methods they support and could document, none of that would be of concern to Solid-Node-Client

**Comment by CxRes** [comment]
I told you before I am thinking of file login in terms of a solid-os, Lets take a random protocol otherwise for now, say ftp. Credential issues will still apply.

**Comment by CxRes** [comment]
> c.fileLogin(fileCred)

Then you would also have to do c.ftpLoigin(cred) etc. I don't think that would be a good idea...

**Comment by jeff-zucker** [comment]
Hmmm.  so login(credentials,protocol='https')  to default to https but able to set it for something else, I like that.

**Comment by CxRes** [comment]
Do you mean
```js
const credentials = {
  ...
  protocol: 'https',
};

session = await c.login(credentials);
```
is the same as:

```js
const credentials = {
  ...
};

session = await c.login(credentials);
```

**Comment by CxRes** [comment]
I was thinking more like: 
```js
const credentials = [{c1}, {c2}, ...];
 
session = await c.login(credentials);
```

**Comment by jeff-zucker** [comment]
```javascript
const nssCreds = {...};
const ftpCreds = {...}'
let session1 = login( nssCreds ) // standard https login
let session2 = login( ftpCreds, 'ftp') // othe login 
```

**Comment by CxRes** [comment]
That works too!!!!

**Comment by jeff-zucker** [comment]
I have it working now with no requires, default import of solid-rest & node-fetch, conditional import on login for the auth packages

**Comment by jeff-zucker** [comment]
I think you were right that I don't need a user-defined fallback - they should just override http which will default to node-fetch

**Comment by CxRes** [comment]
I wasn't anti fallback but not thrilled with fallback including http by default. We do need user defined fallback, in case user wants to handle all other protocols differently (such as throw custom errors).

**Comment by jeff-zucker** [comment]
Yes, user can define fallback, but none is defined by default

**Comment by jeff-zucker** [comment]
Other than that http defaults to node-fetch and file to Solid-Rest.

**Comment by jeff-zucker** [comment]
And unauthenticated https to node-fetch too.

**Comment by CxRes** [comment]
I had a really nice idea regarding logins based on your function signature:

We can have user define multiple logins for the same client, which can be matched against url based on specificity.

Say, we have:
```js
c.login(c1);
c.login(c2, ['https://www.foo.com', 'https://bar.com/private']);
```

A fetch request to `https://www.foo.com/something` and `https://www.bar.com/private/something` will go use `c2` aand other `https://` requests will use `c1`. This way we can restrict the servers to which we send auth tokens.

This I think also solves the issue of login on local networks, since I can use different credentials per IP address or even for specific files on same PC (but accessible to different accounts). 

I would love it if a security expert can look at this!

**Comment by jeff-zucker** [comment]
No, a client can only be logged into one place using a given protocol at  once.  You can have multiple clients though, each one logged in to a different space.  So client a could be logged in to an https site and an ftp site and fetch appropriately from either one, but if you want to go to  different site with the same protocol (e.g. another https site with diff credentials), you either need to logout of that client or start a second one.

**Comment by jeff-zucker** [comment]
So now I am on to rationalizing solid-rest an its plugins.  I am thinking of a repo like this:
```
solid-rest
     package.json
     src
    plugins
        solid-rest-file
            package.json
            src
         solid-rest-mem
            package.json
            src
```
So I  can maintain everything in one repo but the three packages would each have their own dependencies.  An npm install would install all of the package but only the dependencies of solid-rest.  If user wants to use a given plugin, they need to cd into its folder and npm install which would then install its dependencies.

**Comment by jeff-zucker** [comment]
Dude! I am so glad you ignored my grump "why is this important now?" and pressed on until we came up with a login solution.  Since we had the interface, I couldn't resist trying to build something to use it.  I now have solid-rest-ssh working for read and soon write!  Yes, I can access any ssh-addressable host using Solid methods, same as with file!  Thanks for perservering!

**Comment by CxRes** [comment]
> No, a client can only be logged into one place using a given protocol at once. 

Why? Is there a technical reason for this? I see utility of clients with multiple logins as well as having multiple clients! 

> rationalizing solid-rest an its plugins...

The layout is fine.

I have a suggestion:
+ Use pnpm instead of npm and take advantage of their recursive functions and workspace protocols. I have used it for a couple of months and really like the dev experience. Its lighter than using something like Lerna (and better than npm or yarn) but powerful enough to manage a simple monorepo  well. Play with it for half a day, I am sure you will like it!
(What I am not sure of is using actual monorepo or polyrepo when hosting the code on github, I would like to know your experience.)
+ You can integrate Changesets or even Rush (if you ever need a heavy duty solution later).
+ Create a doc folder that github automatically interprets.
+ **Please add a linter**

**Comment by jeff-zucker** [comment]
I am involved in the SolidOS project (rdflib,mashlib,etc.) and it uses lerna.  I haven't even looked at it yet, but I pretty much need to use the same tools as others in the project.

Okay, I guess I can figure out how to add a linter, never done that.

I need to think about he multiple logins thing.  if a single login is usable on multiple sites, I don't need to repeatedly login or specify those sites, I just login once and use them  Your way seems like a nightmare to document and would require storing a hash of logins/urlPatterns and checking them on every fetch.  In terms of restricting servers we send tokens to - just create a client and don't login anywhere with it and it will never send tokens.
